[BUG] Integer Overflow in Kudo when batch size is above max_integer

**Describe the bug**
The following (job fatal) failure is seen when I set batch size as 5 GB for the query (shown below). When I turn `spark.conf.set("spark.rapids.shuffle.kudo.serializer.enabled", "false")` I no longer see this issue. I added a debug log in `GpuPartitioning.scala` to print the sizes around the code where it decides the split the data or not. (also shown below)
Stack trace: 
```
25/08/04 16:05:51 WARN TaskSetManager: Lost task 11.3 in stage 6.0 (TID 4838) (10.185.103.31 executor 6): TaskKilled (Stage cancelled: Job aborted due to stage failure: Task 35 in stage 6.0 failed 4 times, most recent failure: Lost task 35.3 in stage 6.0 (TID 4831) (10.185.103.26 executor 7): java.lang.RuntimeException: java.lang.ArithmeticException: integer overflow
	at com.nvidia.spark.rapids.jni.kudo.KudoSerializer.mergeOnHost(KudoSerializer.java:374)
	at com.nvidia.spark.rapids.KudoTableOperator.$anonfun$concatOnHost$7(GpuShuffleCoalesceExec.scala:293)
	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:57)
	at com.nvidia.spark.rapids.KudoTableOperator.concatOnHost(GpuShuffleCoalesceExec.scala:292)
	at com.nvidia.spark.rapids.KudoTableOperator.concatOnHost(GpuShuffleCoalesceExec.scala:259)
	at com.nvidia.spark.rapids.HostCoalesceIteratorBase.$anonfun$concatenateTablesInHost$7(GpuShuffleCoalesceExec.scala:387)
	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$AutoCloseableAttemptSpliterator.next(RmmRapidsRetryIterator.scala:537)
	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$RmmRapidsRetryIterator.next(RmmRapidsRetryIterator.scala:690)
	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$RmmRapidsRetryAutoCloseableIterator.next(RmmRapidsRetryIterator.scala:577)
	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$.drainSingleWithVerification(RmmRapidsRetryIterator.scala:294)
	at com.nvidia.spark.rapids.RmmRapidsRetryIterator$.withRetryNoSplit(RmmRapidsRetryIterator.scala:167)
	at com.nvidia.spark.rapids.HostCoalesceIteratorBase.$anonfun$concatenateTablesInHost$1(GpuShuffleCoalesceExec.scala:386)
	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
	at com.nvidia.spark.rapids.HostCoalesceIteratorBase.concatenateTablesInHost(GpuShuffleCoalesceExec.scala:343)
	at com.nvidia.spark.rapids.HostCoalesceIteratorBase.$anonfun$next$5(GpuShuffleCoalesceExec.scala:448)
	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
	at com.nvidia.spark.rapids.HostCoalesceIteratorBase.next(GpuShuffleCoalesceExec.scala:447)
	at com.nvidia.spark.rapids.HostCoalesceIteratorBase.next(GpuShuffleCoalesceExec.scala:306)
	at com.nvidia.spark.rapids.CloseableBufferedIterator.next(CloseableBufferedIterator.scala:65)
	at com.nvidia.spark.rapids.CloseableBufferedIterator.head(CloseableBufferedIterator.scala:47)
	at com.nvidia.spark.rapids.CloseableBufferedIterator.headOption(CloseableBufferedIterator.scala:54)
	at com.nvidia.spark.rapids.GpuShuffleCoalesceUtils$.$anonfun$getGpuShuffleCoalesceIterator$1(GpuShuffleCoalesceExec.scala:165)
	at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:30)
	at com.nvidia.spark.rapids.GpuShuffleCoalesceUtils$.getGpuShuffleCoalesceIterator(GpuShuffleCoalesceExec.scala:163)
	at com.nvidia.spark.rapids.GpuShuffledSizedHashJoinExec$HostHostUnspillableJoinSizer.setupForJoin(GpuShuffledSizedHashJoinExec.scala:276)
	at com.nvidia.spark.rapids.GpuShuffledSizedHashJoinExec$HostHostUnspillableJoinSizer.setupForJoin$(GpuShuffledSizedHashJoinExec.scala:263)
	at com.nvidia.spark.rapids.GpuShuffledAsymmetricHashJoinExec$HostHostAsymmetricJoinSizer.setupForJoin(GpuShuffledSizedHashJoinExec.scala:1042)
	at com.nvidia.spark.rapids.GpuShuffledAsymmetricHashJoinExec$AsymmetricJoinSizer.getJoinInfoSmallBuildSide(GpuShuffledSizedHashJoinExec.scala:870)
	at com.nvidia.spark.rapids.GpuShuffledAsymmetricHashJoinExec$AsymmetricJoinSizer.getJoinInfo(GpuShuffledSizedHashJoinExec.scala:833)
	at com.nvidia.spark.rapids.GpuShuffledAsymmetricHashJoinExec$AsymmetricJoinSizer.getJoinInfo$(GpuShuffledSizedHashJoinExec.scala:800)
	at com.nvidia.spark.rapids.GpuShuffledAsymmetricHashJoinExec$HostHostAsymmetricJoinSizer.getJoinInfo(GpuShuffledSizedHashJoinExec.scala:1042)
	at com.nvidia.spark.rapids.GpuShuffledSizedHashJoinExec.getHostHostJoinInfo(GpuShuffledSizedHashJoinExec.scala:538)
	at com.nvidia.spark.rapids.GpuShuffledSizedHashJoinExec.$anonfun$internalDoExecuteColumnar$1(GpuShuffledSizedHashJoinExec.scala:423)
	at org.apache.spark.rdd.ZippedPartitionsRDD2.compute(ZippedPartitionsRDD.scala:89)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:367)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:331)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:367)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:331)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:367)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:331)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:367)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:331)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:367)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:331)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:93)
	at org.apache.spark.TaskContext.runTaskWithListeners(TaskContext.scala:166)
	at org.apache.spark.scheduler.Task.run(Task.scala:141)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$4(Executor.scala:621)
	at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally(SparkErrorUtils.scala:64)
	at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally$(SparkErrorUtils.scala:61)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:94)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:624)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
Caused by: java.lang.ArithmeticException: integer overflow
	at java.lang.Math.toIntExact(Math.java:1011)
	at com.nvidia.spark.rapids.jni.kudo.KudoTableMerger.deserializeValidityBuffer(KudoTableMerger.java:205)
	at com.nvidia.spark.rapids.jni.kudo.KudoTableMerger.visit(KudoTableMerger.java:164)
	at com.nvidia.spark.rapids.jni.schema.Visitors.visitSchemaInner(Visitors.java:81)
	at com.nvidia.spark.rapids.jni.schema.Visitors.visitSchema(Visitors.java:61)
	at com.nvidia.spark.rapids.jni.kudo.KudoTableMerger.lambda$merge$0(KudoTableMerger.java:352)
	at com.nvidia.spark.rapids.jni.Arms.closeIfException(Arms.java:33)
	at com.nvidia.spark.rapids.jni.kudo.KudoTableMerger.merge(KudoTableMerger.java:348)
	at com.nvidia.spark.rapids.jni.kudo.KudoSerializer.mergeOnHost(KudoSerializer.java:350)
	at com.nvidia.spark.rapids.jni.kudo.KudoSerializer.mergeOnHost(KudoSerializer.java:369)

```

**Steps/Code to reproduce bug**
I am using sf= 3k NDS data from `catalog_sales` and `web_sales` . Max partition bytes = 512m, number of shuffle partitions is 50. 
```
import org.apache.spark.sql.functions._
spark.conf.set("spark.rapids.sql.enabled", "true")
spark.conf.set("spark.sql.files.maxPartitionBytes", "512m") 
//spark.conf.set("spark.rapids.shuffle.kudo.serializer.enabled", "false")
println(s"Set spark.sql.files.maxPartitionBytes to: ${spark.conf.get("spark.sql.files.maxPartitionBytes")}")
spark.conf.set("spark.sql.shuffle.partitions", "50")
println(s"Set spark.sql.shuffle.partitions to: ${spark.conf.get("spark.sql.shuffle.partitions")}")

val catalogSalesPath = "hdfs:///mypath/orc_sf3k_dec/catalog_sales"
val webSalesPath = "hdfs:///mypath/orc_sf3k_dec/web_sales"
val catalogSalesDF = spark.read.orc(catalogSalesPath)
val webSalesDF = spark.read.orc(webSalesPath)

val filteredCatalogSalesDF = catalogSalesDF.filter(col("cs_bill_customer_sk") % 3 === 0)
val filteredWebSalesDF = webSalesDF.filter(col("ws_bill_customer_sk") % 3 === 0)


val joinCondition = (filteredCatalogSalesDF("cs_bill_customer_sk") === filteredWebSalesDF("ws_bill_customer_sk"))
val joinedDF = filteredCatalogSalesDF.join(filteredWebSalesDF, joinCondition, "left_outer")
val resultDF = joinedDF.drop(filteredWebSalesDF("ws_bill_customer_sk")).drop(filteredWebSalesDF("ws_item_sk"))
resultDF.printSchema()
resultDF.write.format("orc").mode("overwrite").save("hdfs:///myout")
println("Job finished. Output written to hdfs:///myout/")
```

Log line from `GpuPartitioning.scala`
```
25/08/04 16:05:06 WARN GpuHashPartitioning: mightNeedToSplit=false, maxCpuBatchSize=2147481591,totalInputSize=341028128, hostPartColumns=33,numRows=2516808
```
based on : 
```
logWarning(s"mightNeedToSplit=${mightNeedToSplit}, maxCpuBatchSize=${maxCpuBatchSize}," +
        s"totalInputSize=${totalInputSize}, hostPartColumns=${hostPartColumns.length}," +
        s"numRows=${numRows}")
```
**Expected behavior**
Not fail and split the data
**Environment details (please complete the following information)**
 - Standalone, Spark rapids 25.10 snapshot. Last commit : 2bc01481ea31a22db992a2ef30b2a46de5367a33

--- a/src/main/java/com/nvidia/spark/rapids/jni/kudo/ColumnViewInfo.java
+++ b/src/main/java/com/nvidia/spark/rapids/jni/kudo/ColumnViewInfo.java
@@ -19,7 +19,6 @@ package com.nvidia.spark.rapids.jni.kudo;
 import ai.rapids.cudf.*;
 
 import java.util.Optional;
-import java.util.OptionalLong;
 
 import static com.nvidia.spark.rapids.jni.Preconditions.ensureNonNegative;
 

--- a/src/main/java/com/nvidia/spark/rapids/jni/kudo/KudoTableMerger.java
+++ b/src/main/java/com/nvidia/spark/rapids/jni/kudo/KudoTableMerger.java
@@ -49,7 +49,7 @@ class KudoTableMerger implements SimpleSchemaVisitor {
 
   private final KudoTable[] kudoTables;
   private final ColumnOffsetInfo[] columnOffsets;
-  private final long[] rowCounts;
+  private final int[] rowCounts;
   private final HostMemoryBuffer buffer;
   private final ColumnViewInfo[] colViewInfoList;
   private final long[] validityOffsets;
@@ -67,7 +67,7 @@ class KudoTableMerger implements SimpleSchemaVisitor {
 
   public KudoTableMerger(KudoTable[] tables, HostMemoryBuffer buffer,
                          ColumnOffsetInfo[] columnOffsets,
-                         long[] rowCounts) {
+                         int[] rowCounts) {
     this.kudoTables = requireNonNull(tables, "tables can't be null");
     requireNonNull(buffer, "buffer can't be null!");
     ensure(columnOffsets != null, "column offsets cannot be null");
@@ -108,9 +108,9 @@ class KudoTableMerger implements SimpleSchemaVisitor {
   public void preVisitStruct(Schema structType) {
     ColumnOffsetInfo offsetInfo = getCurColumnOffsets();
     long nullCount = deserializeValidityBuffer(offsetInfo);
-    long totalRowCount = rowCounts[curColIdx];
+    int totalRowCount = rowCounts[curColIdx];
     colViewInfoList[curColIdx] = new ColumnViewInfo(structType.getType(),
-        offsetInfo, toIntExact(nullCount), toIntExact(totalRowCount));
+        offsetInfo, toIntExact(nullCount), totalRowCount);
 
 
     for (int i=0; i<kudoTables.length; i++) {
@@ -131,11 +131,11 @@ class KudoTableMerger implements SimpleSchemaVisitor {
   public void preVisitList(Schema listType) {
     ColumnOffsetInfo offsetInfo = getCurColumnOffsets();
     long nullCount = deserializeValidityBuffer(offsetInfo);
-    long totalRowCount = rowCounts[curColIdx];
+    int totalRowCount = rowCounts[curColIdx];
     deserializeOffsetBuffer(offsetInfo);
 
     colViewInfoList[curColIdx] = new ColumnViewInfo(listType.getType(),
-        offsetInfo, toIntExact(nullCount), toIntExact(totalRowCount));
+        offsetInfo, toIntExact(nullCount), totalRowCount);
 
     for (int i=0; i<kudoTables.length; i++) {
       KudoTableHeader header = kudoTables[i].getHeader();
@@ -162,7 +162,7 @@ class KudoTableMerger implements SimpleSchemaVisitor {
   public void visit(Schema primitiveType) {
     ColumnOffsetInfo offsetInfo = getCurColumnOffsets();
     long nullCount = deserializeValidityBuffer(offsetInfo);
-    long totalRowCount = rowCounts[curColIdx];
+    int totalRowCount = rowCounts[curColIdx];
     if (primitiveType.getType().hasOffsets()) {
       deserializeOffsetBuffer(offsetInfo);
       deserializeDataBuffer(offsetInfo, OptionalInt.empty());
@@ -171,7 +171,7 @@ class KudoTableMerger implements SimpleSchemaVisitor {
     }
 
     colViewInfoList[curColIdx] = new ColumnViewInfo(primitiveType.getType(),
-        offsetInfo, toIntExact(nullCount), toIntExact(totalRowCount));
+        offsetInfo, toIntExact(nullCount), totalRowCount);
 
     if (primitiveType.getType().hasOffsets()) {
       for (int i=0; i<kudoTables.length; i++) {

--- a/src/main/java/com/nvidia/spark/rapids/jni/kudo/MergedInfoCalc.java
+++ b/src/main/java/com/nvidia/spark/rapids/jni/kudo/MergedInfoCalc.java
@@ -36,7 +36,7 @@ class MergedInfoCalc implements SimpleSchemaVisitor {
     // Total data len in gpu, which accounts for 64 byte alignment
     private long totalDataLen;
     private final boolean[] hasNull;
-    private final long[] rowCount;
+    private final int[] rowCount;
     private final long[] dataLen;
 
     // Column offset in gpu device buffer, it has one field for each flattened column
@@ -50,7 +50,7 @@ class MergedInfoCalc implements SimpleSchemaVisitor {
         int columnCount = tables[0].getHeader().getNumColumns();
         this.hasNull = new boolean[columnCount];
         initHasNull();
-        this.rowCount = new long[columnCount];
+        this.rowCount = new int[columnCount];
         this.dataLen = new long[columnCount];
         this.columnOffsets = new ColumnOffsetInfo[columnCount];
     }
@@ -73,7 +73,7 @@ class MergedInfoCalc implements SimpleSchemaVisitor {
         return kudoTables;
     }
 
-    public long[] getRowCount() {
+    public int[] getRowCount() {
         return rowCount;
     }
 
@@ -228,7 +228,7 @@ class MergedInfoCalc implements SimpleSchemaVisitor {
         @Override
         public void preVisitStruct(Schema structType) {
             SliceInfo sliceInfo = sliceInfos.getLast();
-            rowCount[curColIdx] += sliceInfo.getRowCount();
+            rowCount[curColIdx] = Math.addExact(rowCount[curColIdx], sliceInfo.getRowCount());
 
             curColIdx++;
         }
@@ -241,7 +241,7 @@ class MergedInfoCalc implements SimpleSchemaVisitor {
         @Override
         public void preVisitList(Schema listType) {
             SliceInfo sliceInfo = sliceInfos.getLast();
-            rowCount[curColIdx] += sliceInfo.getRowCount();
+            rowCount[curColIdx] = Math.addExact(rowCount[curColIdx], sliceInfo.getRowCount());
 
             if (sliceInfo.getRowCount() > 0) {
                 int startOffset = table.getBuffer().getInt(bufferOffset);
@@ -265,7 +265,7 @@ class MergedInfoCalc implements SimpleSchemaVisitor {
         @Override
         public void visit(Schema primitiveType) {
             SliceInfo sliceInfo = sliceInfos.getLast();
-            rowCount[curColIdx] += sliceInfo.getRowCount();
+            rowCount[curColIdx] = Math.addExact(rowCount[curColIdx], sliceInfo.getRowCount());
             if (primitiveType.getType().hasOffsets()) {
                 // string type
                 if (sliceInfo.getRowCount() > 0) {

--- a/src/test/java/com/nvidia/spark/rapids/jni/kudo/KudoSerializerTest.java
+++ b/src/test/java/com/nvidia/spark/rapids/jni/kudo/KudoSerializerTest.java
@@ -694,9 +694,9 @@ public class KudoSerializerTest extends CudfTestBase {
 
   @Test
   public void testSerializeAndDeserializeLargeTable() {
-    try(Table t1 = buildLargeTestTable()) {
-      try (Table t2 = buildLargeTestTable()) {
-        try (Table expected = Table.concatenate(t1, t2)) {
+    try(Table t1 = buildLargeTestTable();
+        Table t2 = buildLargeTestTable();
+        Table expected = Table.concatenate(t1, t2)) {
           assertTrue(expected.getDeviceMemorySize() > Integer.MAX_VALUE,
               "Expected table size should exceed Integer.MAX_VALUE");
           final int sliceSize = 50000000;
@@ -714,8 +714,6 @@ public class KudoSerializerTest extends CudfTestBase {
           }
 
           checkMergeTable(expected, tableSlices);
-        }
-      }
     } catch (Exception e) {
       throw new RuntimeException(e);
     }

